models/repo_indexer.go

```diff
@@ -0,0 +1,294 @@
+// Copyright 2017 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"io/ioutil"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+
+	"code.gitea.io/git"
+	"code.gitea.io/gitea/modules/base"
+	"code.gitea.io/gitea/modules/indexer"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+
+	"github.com/Unknwon/com"
+)
+
+// RepoIndexerStatus status of a repo's entry in the repo indexer
+// For now, implicitly refers to default branch
+type RepoIndexerStatus struct {
+	ID        int64  `xorm:"pk autoincr"`
+	RepoID    int64  `xorm:"INDEX"`
+	CommitSha string `xorm:"VARCHAR(40)"`
+}
+
+func (repo *Repository) getIndexerStatus() error {
+	if repo.IndexerStatus != nil {
+		return nil
+	}
+	status := &RepoIndexerStatus{RepoID: repo.ID}
+	has, err := x.Get(status)
+	if err != nil {
+		return err
+	} else if !has {
+		status.CommitSha = ""
+	}
+	repo.IndexerStatus = status
+	return nil
+}
+
+func (repo *Repository) updateIndexerStatus(sha string) error {
+	if err := repo.getIndexerStatus(); err != nil {
+		return err
+	}
+	if len(repo.IndexerStatus.CommitSha) == 0 {
+		repo.IndexerStatus.CommitSha = sha
+		_, err := x.Insert(repo.IndexerStatus)
+		return err
+	}
+	repo.IndexerStatus.CommitSha = sha
+	_, err := x.ID(repo.IndexerStatus.ID).Cols("commit_sha").
+		Update(repo.IndexerStatus)
```

Comment:
Why it's own table if we're using it as a field? Just put `LastIndexedSHA` in the `Repository` table instead?

Reply:
@bkcsoft that was lunny request for this so that it could be possible to implement search on multiple branches in future

Reply:
@lafriks @lunny Makes the PR harder to read though :(

Reply:
:(